src/adapters/bedrock/chat.ts

```diff
@@ -0,0 +1,289 @@
+        new Client({
+          region: value.config.region,
+          credentials: value.config.credentials,
+        }),
+    });
+  }
+
+  async meta(): Promise<LLMMeta> {
+    if (this.modelId.includes("titan-text-premier")) {
+      return { tokenLimit: 3 * 1024 };
+    } else if (
+      this.modelId.includes("titan-text-express") ||
+      this.modelId.includes("anthropic.claude-v2") ||
+      this.modelId.includes("anthropic.claude-instant-v1") ||
+      this.modelId.includes("anthropic.claude-3-sonnet") ||
+      this.modelId.includes("anthropic.claude-3-haiku") ||
+      this.modelId.includes("anthropic.claude-3-opus") ||
+      this.modelId.includes("meta.llama2") ||
+      this.modelId.includes("cohere.command-text") ||
+      this.modelId.includes("cohere.command-light")
+    ) {
+      return { tokenLimit: 4 * 1024 };
+    } else if (
+      this.modelId.includes("titan-text-lite") ||
+      this.modelId.includes("anthropic.claude-3-5-sonnet") ||
+      this.modelId.includes("anthropic.claude-3-5-haiku") ||
+      this.modelId.includes("meta.llama3-8b") ||
+      this.modelId.includes("meta.llama3-70b") ||
+      this.modelId.includes("ai21.j2")
+    ) {
+      return { tokenLimit: 8 * 1024 };
+    } else if (
+      this.modelId.includes("mistral.mistral-7b") ||
+      this.modelId.includes("mistral.mixtral-8x7b") ||
+      this.modelId.includes("mistral.mistral-small")
+    ) {
+      return { tokenLimit: 32 * 1024 };
+    } else if (
+      this.modelId.includes("meta.llama3-1") ||
+      this.modelId.includes("meta.llama3-2") ||
+      this.modelId.includes("mistral.mistral-large") ||
+      this.modelId.includes("cohere.command-r")
+    ) {
+      return { tokenLimit: 128 * 1024 };
+    } else if (this.modelId.includes("ai21.jamba")) {
+      return { tokenLimit: 256 * 1024 };
+    }
+
+    return {
+      tokenLimit: Infinity,
+    };
+  }
+
+  async tokenize(input: BaseMessage[]): Promise<BaseLLMTokenizeOutput> {
+    const contentLength = input.reduce((acc, msg) => acc + msg.text.length, 0);
+    return { tokensCount: Math.ceil(contentLength / 4) };
+  }
+
+  protected convertToConverseMessages(messages: BaseMessage[]): {
+    conversation: BedrockMessage[];
+    systemMessage: BedrockSystemContentBlock[];
+  } {
+    const systemMessage: BedrockSystemContentBlock[] = messages
+      .filter((msg) => msg.role === Role.SYSTEM)
+      .map((msg) => {
+        return { text: msg.text };
+      });
+    const converseMessages: BedrockMessage[] = messages
+      .filter((msg) => msg.role !== Role.SYSTEM)
+      .map((msg) => {
+        return {
+          role: msg.role === Role.USER ? Role.USER : Role.ASSISTANT,
+          content: [{ text: msg.text }],
+        };
+      });
+
+    const conversation = converseMessages.reduce<BedrockMessage[]>(
+      (messageList, currentMessage) => {
+        const lastMessage = messageList[messageList.length - 1];
+        if (lastMessage && lastMessage.role === Role.USER) {
```

Comment:
What if the `lastMessage === currentMessage`? 
In that case, you would duplicate the message, which does not seem right.

Reply:
In the reduce loop, lastMessage is always the last item in the messageList (accumulator), and currentMessage is the next item being processed. Unless the same object (i.e., due to a shared reference) is repeated in the input, lastMessage and currentMessage will always be distinct. While I tested the code with and without the lastMessage !== currentMessage condition and found no difference in behaviour, I will add the condition as a safety net.